ex19/doc/Wheel.3
'\" t
.TH "" 3 "" "Version Unknown To Mankind" "Free Widget Foundation"
.SH NAME
Wheel
.SH DESCRIPTION
simple Wheel Input widget




.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Wheel
Name	Class	Type	Default
XtNxftFont	XtCXFtFont	XftFont	"Sans-22"
XtNcallback	XtCCallback	Callback	NULL 
XtNbg_norm	XtCBg_norm	Pixel	"lightblue"
XtNbg_sel	XtCBg_sel	Pixel	"yellow"
XtNbg_hi	XtCBg_hi	Pixel	"red"
XtNfg_norm	XtCFg_norm	Pixel	"black"
XtNfg_sel	XtCFg_sel	Pixel	"green"
XtNfg_hi	XtCFg_hi	Pixel	"white"
XtNuser_data	XtCUser_data	int 	0 
XtNfocus_group	XtCFocus_group	String 	""
XtNstate	XtCState	int 	0 
XtNregister_focus_group	XtCRegister_focus_group	Boolean 	True 

.TE
.ps +2


.TP
.I "XtNxftFont"



.TP
.I "XtNcallback"



.TP
.I "XtNbg_norm"



.TP
.I "XtNbg_sel"



.TP
.I "XtNbg_hi"



.TP
.I "XtNfg_norm"



.TP
.I "XtNfg_sel"



.TP
.I "XtNfg_hi"



.TP
.I "XtNuser_data"



.TP
.I "XtNfocus_group"



.TP
.I "XtNstate"



.TP
.I "XtNregister_focus_group"



.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2

.SS "Exports"


.nf

.B type
 wheel_state = enum {
  STATE_NORMAL = 0,
  STATE_SELECTED = 1,
  STATE_ARMED = 2,
  STATE_MAX = 3
}
.fi



.nf

.B type
 wheel_cmd = enum {
      WHEEL_NOCOMMAND,
      WHEEL_LEFT,
      WHEEL_RIGHT,
      WHEEL_FIRE,
      WHEEL_KEYPAD
}
.fi



.nf

.B type
 wheel_colors = enum {
 COLOR_FG_NORM, COLOR_FG_SEL, COLOR_FG_HI, COLOR_BG_NORM, COLOR_BG_SEL, COLOR_BG_HI
}
.fi


\fBdef\fP COLOR_FG_MAX = 3 


.nf

.B type
 wheel_gcs = enum {
 GC_BG_NORM, GC_BG_SEL, GC_BG_HI
}
.fi



.nf
int  wheel_exec_command( $, int  cmd, int  val)
.fi



.nf
int  wheel_sig_recv( $, Widget  from, int  type, void * class, void * data)
.fi



ex19/src/Wheel.c
/* Generated by wbuild
 * (generator version 3.3)
 */
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 181 "../wb/Wheel.widget"
#include <X11/Xft/Xft.h>
#line 182 "../wb/Wheel.widget"
#include "converters-xft.h"
#line 183 "../wb/Wheel.widget"
#include "focus-group.h"
#line 184 "../wb/Wheel.widget"
#include "apputil.h"
#line 185 "../wb/Wheel.widget"
#include "xutil.h"
#include <xtcw/WheelP.h>
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 83 "../wb/Wheel.widget"
static void class_initialize(
#if NeedFunctionPrototypes
void
#endif
);
#line 85 "../wb/Wheel.widget"
static int  exec_command(
#if NeedFunctionPrototypes
Widget,int ,int 
#endif
);
#line 91 "../wb/Wheel.widget"
static int  sig_recv(
#if NeedFunctionPrototypes
Widget,Widget ,int ,void *,void *
#endif
);
#line 98 "../wb/Wheel.widget"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 119 "../wb/Wheel.widget"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 136 "../wb/Wheel.widget"
static void destroy(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 150 "../wb/Wheel.widget"
static void col_set(
#if NeedFunctionPrototypes
Widget,int 
#endif
);
#line 150 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 150 "../wb/Wheel.widget"
static void col_set(Widget self,int  num)
#else
#line 150 "../wb/Wheel.widget"
static void col_set(self,num)Widget self;int  num;
#endif
#line 151 "../wb/Wheel.widget"
{
    Display *dpy = XtDisplay(self);
    XGCValues     values;

    switch( num ) {
    case COLOR_FG_NORM: ((WheelWidget)self)->wheel.pixel[COLOR_FG_NORM]= ((WheelWidget)self)->wheel.fg_norm; break;
    case COLOR_FG_SEL: ((WheelWidget)self)->wheel.pixel[COLOR_FG_SEL]= ((WheelWidget)self)->wheel.fg_sel; break;
    case COLOR_FG_HI: ((WheelWidget)self)->wheel.pixel[COLOR_FG_HI]= ((WheelWidget)self)->wheel.fg_hi; break;
    case COLOR_BG_NORM: ((WheelWidget)self)->wheel.pixel[COLOR_BG_NORM]= ((WheelWidget)self)->wheel.bg_norm; break;
    case COLOR_BG_SEL: ((WheelWidget)self)->wheel.pixel[COLOR_BG_SEL]= ((WheelWidget)self)->wheel.bg_sel; break;
    case COLOR_BG_HI: ((WheelWidget)self)->wheel.pixel[COLOR_BG_HI]= ((WheelWidget)self)->wheel.bg_hi; break;
    }

    /* !TODO! Check if Color is already alloced
      if( $xft_col[num] )
        XftColorFree(dpy, DefaultVisual(dpy, DefaultScreen(dpy)),
		 None, $xft_col+num);
    */
    xft_color_alloc( self, ((WheelWidget)self)->wheel.pixel[num], ((WheelWidget)self)->wheel.xft_col+num );

    if( num >= 3 ) {
        values.foreground = ((WheelWidget)self)->wheel.pixel[num]; /* background color */
        if( ((WheelWidget)self)->wheel.gc[num-3] ) XtReleaseGC(self,((WheelWidget)self)->wheel.gc[num-3]);
        ((WheelWidget)self)->wheel.gc[num-3] = XtGetGC(self, GCForeground, &values);
    }
}

static XtResource resources[] = {
#line 7 "../wb/Wheel.widget"
{XtNxftFont,XtCXFtFont,XtRXftFont,sizeof(((WheelRec*)NULL)->wheel.xftFont),XtOffsetOf(WheelRec,wheel.xftFont),XtRString,(XtPointer)"Sans-22"},
#line 8 "../wb/Wheel.widget"
{XtNcallback,XtCCallback,XtRCallback,sizeof(((WheelRec*)NULL)->wheel.callback),XtOffsetOf(WheelRec,wheel.callback),XtRImmediate,(XtPointer)NULL },
#line 9 "../wb/Wheel.widget"
{XtNbg_norm,XtCBg_norm,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.bg_norm),XtOffsetOf(WheelRec,wheel.bg_norm),XtRString,(XtPointer)"lightblue"},
#line 10 "../wb/Wheel.widget"
{XtNbg_sel,XtCBg_sel,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.bg_sel),XtOffsetOf(WheelRec,wheel.bg_sel),XtRString,(XtPointer)"yellow"},
#line 11 "../wb/Wheel.widget"
{XtNbg_hi,XtCBg_hi,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.bg_hi),XtOffsetOf(WheelRec,wheel.bg_hi),XtRString,(XtPointer)"red"},
#line 12 "../wb/Wheel.widget"
{XtNfg_norm,XtCFg_norm,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.fg_norm),XtOffsetOf(WheelRec,wheel.fg_norm),XtRString,(XtPointer)"black"},
#line 13 "../wb/Wheel.widget"
{XtNfg_sel,XtCFg_sel,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.fg_sel),XtOffsetOf(WheelRec,wheel.fg_sel),XtRString,(XtPointer)"green"},
#line 14 "../wb/Wheel.widget"
{XtNfg_hi,XtCFg_hi,XtRPixel,sizeof(((WheelRec*)NULL)->wheel.fg_hi),XtOffsetOf(WheelRec,wheel.fg_hi),XtRString,(XtPointer)"white"},
#line 15 "../wb/Wheel.widget"
{XtNuser_data,XtCUser_data,XtRInt,sizeof(((WheelRec*)NULL)->wheel.user_data),XtOffsetOf(WheelRec,wheel.user_data),XtRImmediate,(XtPointer)0 },
#line 16 "../wb/Wheel.widget"
{XtNfocus_group,XtCFocus_group,XtRString,sizeof(((WheelRec*)NULL)->wheel.focus_group),XtOffsetOf(WheelRec,wheel.focus_group),XtRString,(XtPointer)""},
#line 17 "../wb/Wheel.widget"
{XtNstate,XtCState,XtRInt,sizeof(((WheelRec*)NULL)->wheel.state),XtOffsetOf(WheelRec,wheel.state),XtRImmediate,(XtPointer)0 },
#line 18 "../wb/Wheel.widget"
{XtNregister_focus_group,XtCRegister_focus_group,XtRBoolean,sizeof(((WheelRec*)NULL)->wheel.register_focus_group),XtOffsetOf(WheelRec,wheel.register_focus_group),XtRImmediate,(XtPointer)True },
};

WheelClassRec wheelClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &coreClassRec,
/* class_name   	*/  "Wheel",
/* widget_size  	*/  sizeof(WheelRec),
/* class_initialize 	*/  class_initialize,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  XtInheritRealize,
/* actions      	*/  NULL,
/* num_actions  	*/  0,
/* resources    	*/  resources,
/* num_resources 	*/  12,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  False ,
/* compress_exposure 	*/  FALSE ,
/* compress_enterleave 	*/  False ,
/* visible_interest 	*/  False ,
/* destroy      	*/  destroy,
/* resize       	*/  XtInheritResize,
/* expose       	*/  XtInheritExpose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  NULL,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* Wheel_class part */
exec_command,
sig_recv,
},
};
WidgetClass wheelWidgetClass = (WidgetClass) &wheelClassRec;
static void _resolve_inheritance(class)
WidgetClass class;
{
  WheelWidgetClass c __attribute__((unused)) = (WheelWidgetClass) class;
  WheelWidgetClass super __attribute__((unused));
  if (class == wheelWidgetClass) return;
  super = (WheelWidgetClass)class->core_class.superclass;
  if (c->wheel_class.exec_command == XtInherit_exec_command)
    c->wheel_class.exec_command = super->wheel_class.exec_command;
  if (c->wheel_class.sig_recv == XtInherit_sig_recv)
    c->wheel_class.sig_recv = super->wheel_class.sig_recv;
}
#line 83 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 83 "../wb/Wheel.widget"
static void class_initialize(void)
#else
#line 83 "../wb/Wheel.widget"
static void class_initialize()
#endif
#line 84 "../wb/Wheel.widget"
{ converters_xft_init(); }
#line 85 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 85 "../wb/Wheel.widget"
static int  exec_command(Widget self,int  cmd,int  val)
#else
#line 85 "../wb/Wheel.widget"
static int  exec_command(self,cmd,val)Widget self;int  cmd;int  val;
#endif
#line 86 "../wb/Wheel.widget"
{
      XtWarning( "Wheel Widget needs exec_command to be redefined" );
      return 0;
}
#line 91 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 91 "../wb/Wheel.widget"
static int  sig_recv(Widget self,Widget  from,int  type,void * class,void * data)
#else
#line 91 "../wb/Wheel.widget"
static int  sig_recv(self,from,type,class,data)Widget self;Widget  from;int  type;void * class;void * data;
#endif
#line 92 "../wb/Wheel.widget"
{
  XtWarning( "Wheel Widget needs sig_recv to be redefined" );
  return 0;
}
#line 98 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 98 "../wb/Wheel.widget"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 98 "../wb/Wheel.widget"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 99 "../wb/Wheel.widget"
{
    Bool t = False;
    if( ((WheelWidget)old)->wheel.fg_norm != ((WheelWidget)self)->wheel.fg_norm ) { t=True;
        col_set(self,COLOR_FG_NORM); }
    if( ((WheelWidget)old)->wheel.fg_sel != ((WheelWidget)self)->wheel.fg_sel ) { t=True;
        col_set(self,COLOR_FG_SEL); }
    if( ((WheelWidget)old)->wheel.fg_hi != ((WheelWidget)self)->wheel.fg_hi ) { t=True;
        col_set(self,COLOR_FG_HI); }
    if( ((WheelWidget)old)->wheel.bg_norm != ((WheelWidget)self)->wheel.bg_norm ) { t=True;
        col_set(self,COLOR_BG_NORM); }
    if( ((WheelWidget)old)->wheel.bg_sel != ((WheelWidget)self)->wheel.bg_sel ) { t=True;
        col_set(self,COLOR_BG_SEL); }
    if( ((WheelWidget)old)->wheel.bg_hi != ((WheelWidget)self)->wheel.bg_hi ) { t=True;
        col_set(self,COLOR_BG_HI ); }
    if( ((WheelWidget)old)->wheel.xftFont != ((WheelWidget)self)->wheel.xftFont ) { t=True;
        XftFontClose(XtDisplay(self),((WheelWidget)old)->wheel.xftFont); }
    return t;
}
#line 119 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 119 "../wb/Wheel.widget"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 119 "../wb/Wheel.widget"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 120 "../wb/Wheel.widget"
{
  int i;
  memset( ((WheelWidget)self)->wheel.gc, 0, sizeof( ((WheelWidget)self)->wheel.gc ));
  memset( ((WheelWidget)self)->wheel.xft_col, 0, sizeof( ((WheelWidget)self)->wheel.xft_col));
  for(i=0;i<6;i++)
      col_set(self,i);

  ((WheelWidget)self)->wheel.state = STATE_NORMAL;
  ((WheelWidget)self)->core.background_pixel = ((WheelWidget)self)->wheel.pixel[STATE_NORMAL+3];

  if( ! STR_IS_EMPTY(((WheelWidget)self)->wheel.focus_group) && ((WheelWidget)self)->wheel.register_focus_group ) {
      focus_add2( self, ((WheelWidget)self)->wheel.focus_group );
  }

}
#line 136 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 136 "../wb/Wheel.widget"
static void destroy(Widget self)
#else
#line 136 "../wb/Wheel.widget"
static void destroy(self)Widget self;
#endif
#line 137 "../wb/Wheel.widget"
{
  Display *dpy = XtDisplay(self);

  int i;
  for(i=0;i<6;i++) {
    XftColorFree(dpy, DefaultVisual(dpy, DefaultScreen(dpy)),
		 None, ((WheelWidget)self)->wheel.xft_col+i);
    if( i < 3 ) XtReleaseGC(self,((WheelWidget)self)->wheel.gc[i]);
  }
}
#line 22 "../wb/Wheel.widget"
#line 29 "../wb/Wheel.widget"
#line 37 "../wb/Wheel.widget"
#line 40 "../wb/Wheel.widget"
#line 43 "../wb/Wheel.widget"
#line 48 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 48 "../wb/Wheel.widget"
int  wheel_exec_command(Widget self,int  cmd,int  val)
#else
#line 48 "../wb/Wheel.widget"
int  wheel_exec_command(self,cmd,val)Widget self;int  cmd;int  val;
#endif
#line 49 "../wb/Wheel.widget"
{
  if( XtIsSubclass( self, wheelWidgetClass) && ((WheelWidgetClass)self->core.widget_class)->wheel_class.exec_command )
    {
      return ((WheelWidgetClass)self->core.widget_class)->wheel_class.exec_command(self,cmd,val);
    }

  XtWarning( "Widget is not subclass of wheel" );
  return 0;
}
#line 59 "../wb/Wheel.widget"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 59 "../wb/Wheel.widget"
int  wheel_sig_recv(Widget self,Widget  from,int  type,void * class,void * data)
#else
#line 59 "../wb/Wheel.widget"
int  wheel_sig_recv(self,from,type,class,data)Widget self;Widget  from;int  type;void * class;void * data;
#endif
#line 60 "../wb/Wheel.widget"
{
  if( XtIsSubclass( self, wheelWidgetClass) && ((WheelWidgetClass)self->core.widget_class)->wheel_class.sig_recv )
    {
      return ((WheelWidgetClass)self->core.widget_class)->wheel_class.sig_recv(self,from,type,class,data);
    }

  XtWarning( "Widget is not subclass of wheel" );
  return 0;
}
